Add convolution node

piedoom opened the pull request:
(Draft) Implements #15. Managed to get something working and wanted to start a draft with some notes to get any early thoughts. Feel free to reject entirely if this isn't up to snuff, I'm not the best at this, and is good practice in any case! Appreciate any feedback or comments on performance.

## Still to-do

- Behaving correctly when the IR sample is changed, or IR channel counts are changed.
    - The amount of buffers allocated is dependent on the IR channel count, too, which may change. May need a "maximum" channel support setting so IR channels can change at runtime without allocations.
- Cover cases for when IR sample channels and buffer output channels are not the same count
    - Mono IR + Stereo audio should yield the IR applying to both output channels instead of just the left
- Ensure correctness (or warn/error for users, if preferable) when using unusual input, output
- More robust loading of IR sample in visual node graph example

## Changes with explanation

### Dependencies

This adds the [`fft-convolver`](https://lib.rs/crates/fft-convolver) crate as an optional dependency to `firewheel-nodes`.

### "True Stereo"

The "true stereo" setting for convolution requires the allocation of +2x independent stereo. Because this option is not likely to need to change at runtime, it is a config parameter.

Also because of this, 

### Mono-to-stereo convolution

One potential application of convolution is to add space to a mono audio signal using a stereo IR. I explored adding const generics for both inputs and outputs, but found this overcomplicates things, and instead that users should simply use the mono to stereo node to achieve the same result.

### Other

- I added a CC0 IR asset for testing purposes
- I added `.DS_Store` to `.gitignore` for Mac developers since it was bugging me
- I added convolution nodes of various settings into the visual node graph (and grouped other mono/stereo nodes for a first pass at consistency there, just to reduce noise a bit and improve scan-ability)

Q: Would make sense to add a supertrait for the `num_channels` and `len_frames` definition since both `SampleResource` and `SampleResourceF32` implement it? It could make blanket impls (e.g. on Vecs) a bit more concise.

BillyDM commented:
As for wet/dry mix, I realize I should add wet/dry mix DSP in `firewheel-core` to make that easier. I'll get that done here soon.

(Also don't forget that you will need to delay the dry signal by the same number of frames as the latency of the convolution effect.)

piedoom commented:
Thanks for the comments - it would be nice to have wet/dry in DSP core. I imagine a lot of it is similar to the crossfade node.
Totally forgot about the convolution dry mix delay, will fix that and tidy up some of the other mentioned areas!

piedoom commented:
Ok, got some more progress on this.

1. IR samples can now be changed and will work properly. I added a setting in the config for maximum IR sample channels, defaulting to stereo (But users can select mono to save memory if needed).
2. Mono IRs will fill both channels in a stereo Convolution node. I figure that this is a better assumption for default behavior than just convolving the left output.
3. Switched to using the new MixDSP API
4. Removed "enabled" entirely
5. Added input audio buffer to fix dry timing issue
6. Added supertrait `SampleResourceInfo` and moved `num_channels`, `len_frames` methods into it. LMK if I should put this in a different PR (and also if you have a better name, idk about SampleResourceInfo to be honest)
7. Improved asset loading in the visual graph example and allowed for IR selection in the convolution nodes, including mono versions to test.

- Q: Should this also implement a declick when changing IR samples like the sample player?
- Q: I added an explicit panic if CHANNELS > 2 in the node info. Is this an acceptable way to handle behavior?  
    - Otherwise we will need some sort of mixdown and mixup, assuming that most IRs are stereo, but IMO that should just be another node
    - Or we could not panic but just only ever use the first 2 channels (and maybe warn?)

BillyDM commented:
I think panicking if `CHANNELS > 2` is fine.

piedoom commented:
Alright, I think this is more or less ready. I added a wet amplitude to account for amplitude of convolved samples, and removed true stereo stuff. 

*one* thing I still want to investigate is if the IR buffer can be replaced without re-initializing the convolution buffer (I believe that is why I am hearing clicking when changing samples). I think i might need to open a PR/fork the fft-convolver crate. But yeah, I don't think there is currently a way to mix between two IRs, only two convolution nodes, which isn't ideal.

piedoom commented:
ack, i forgot about the pause/play function mentioned on discord for reverb - i think it is applicable here as well. will push a change with a fix.
